xray/segment.go

```diff
@@ -322,25 +354,27 @@ func (seg *Segment) flush() {
 			logger.Debugf("emit lambda subsegment named: %v", seg.Name)
 			seg.emit()
 		} else {
-			seg.parent.safeFlush()
+			return false
```

Comment:
Looks like you removed the recursiveness to force the caller of `flush()` to handle it. That's probably one of the reasons you have duplicate traverse logic in multiple places. This is not very easy to maintain. I think this particular method wit the new change is the right way. A segment should not flush other segments inside its own locked method. But IMO there should be some central unit to control the whole streaming/emission instead of have a complicated for-loop everywhere on calling `flush()`.

Reply:
Good point, I have refactored logic. A new method `send()` is created which is  thread safe and takes care of sending current (Sub)Segment or parent if its ready to send. Going ahead `send()` should be called.